### PR TITLE
Extend visibility ClusterQueue PendingWorkloads endpoint

### PR DESCRIPTION
**Pull request by PBundyra**
#### What type of PR is this?
/kind feature
/kind api-change

#### What this PR does / why we need it:
This PR continues the development of [KEP#168-2](https://github.com/kubernetes-sigs/kueue/pull/1300)

#### Does this PR introduce a user-facing change?

```release-note
Extend the information returned for the pending workloads in cluster queue, that is used to determine the workload position, including the workload position itself.

```

### CODE REVIEW

File: test/e2e/e2e_test.go [hunk 1]
```diff
@@ -157,6 +161,83 @@ var _ = ginkgo.Describe("Kueue visibility server", func() {
...
+				gomega.Eventually(func() int {
+					var err error
+					info, err = visibilityClient.ClusterQueues().GetPendingWorkloadsSummary(ctx, clusterQueue.Name, metav1.GetOptions{})
+					gomega.Expect(err).NotTo(gomega.HaveOccurred())
+					return len(info.Items)
+				}, util.Timeout, util.Interval).Should(gomega.Equal(3))
+			})
+
+			ginkgo.By("Verify their position in ClusterQueue", func() {
+				gomega.Expect(info.Items[0].PositionInClusterQueue).To(gomega.Equal(0))
```
Comment (mimowo) [inline test/e2e/e2e_test.go:223]: This is hard to read, can we instead use the `ComparableTo` function (or such) to compare the entire list of pending workloads to the expected list?

File: apis/visibility/v1alpha1/types.go [hunk 2]
```diff
@@ -84,10 +84,10 @@ type PendingWorkloadOptions struct {
 	metav1.TypeMeta `json:",inline"`
 
 	// Offset indicates position of the first pending workload that should be fetched starting from 0. 0 by default
-	Offset int32 `json:"offset"`
+	Offset int `json:"offset"`
```
Comment (alculquicondor) [inline apis/visibility/v1alpha1/types.go:87]: use int32 or int64, as necessary.

Comment (PBundyra) [reply apis/visibility/v1alpha1/types.go:87]: int32 or int64 are not supported by the conversion framework and using them would require manual conversion

File: pkg/visibility/api/rest/local_queue.go [hunk 3]
```diff
@@ -0,0 +1,53 @@
+// Copyright 2023 The Kubernetes Authors.
```
Comment (mimowo) [inline pkg/visibility/api/rest/local_queue.go:1]: This seems like unrelated to the PR?

File: pkg/visibility/api/rest/pending_workload_CQ_test.go [hunk 4]
```diff
@@ -292,12 +292,16 @@ func TestPendingWorkloads(t *testing.T) {
...
-				for index, want := range tc.wantPendingWorkloads {
-					got := pendingWorkloadsInfo.Items[index]
-					if diff := cmp.Diff(want, got, cmpopts.IgnoreTypes(metav1.TypeMeta{})); diff != "" {
-						t.Errorf("Pending workload at index %v differ: (-want,+got):\n%s", index, diff)
-					}
+				if diff := cmp.Diff(tc.wantPendingWorkloads, pendingWorkloadsInfo.Items, cmpopts.IgnoreTypes(metav1.TypeMeta{})); diff != "" {
+					t.Errorf("Pending workloads differ: (-want,+got):\n%s", diff)
 				}
+
+				// for index, want := range tc.wantPendingWorkloads {
```
Comment (mimowo) [inline pkg/visibility/api/rest/pending_workload_CQ_test.go:299]: Clear this

File: test/e2e/e2e_test.go [hunk 5]
```diff
@@ -219,22 +222,36 @@ var _ = ginkgo.Describe("Kueue visibility server", func() {
...
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Namespace: ns.Name,
+						},
+						Priority:               lowPriorityClass.Value,
+						PositionInLocalQueue:   1,
+						PositionInClusterQueue: 2,
+					},
+				}
+				gomega.Expect(info.Items).To(gomega.BeComparableTo(expectedWls, cmpopts.IgnoreFields(visibility.PendingWorkload{}, "Name", "LocalQueueName")))
```
Comment (mimowo) [inline test/e2e/e2e_test.go:252]: Why excluding the LocalQueueName here, in favor of the additional step? I think it would read easier to have all the properties here.

File: test/e2e/e2e_test.go [hunk 6]
```diff
@@ -219,22 +222,36 @@ var _ = ginkgo.Describe("Kueue visibility server", func() {
...
-				gomega.Expect(info.Items[2].PositionInClusterQueue).To(gomega.BeNumerically("==", 2))
-			})
-
-			ginkgo.By("Verify two first workloads have high priority and 0 LocalQueue position", func() {
-				gomega.Expect(info.Items[0].Priority).To(gomega.Equal(highPriorityClass.Value))
-				gomega.Expect(info.Items[1].Priority).To(gomega.Equal(highPriorityClass.Value))
-				gomega.Expect(info.Items[0].PositionInLocalQueue).To(gomega.BeNumerically("==", 0))
-				gomega.Expect(info.Items[1].PositionInLocalQueue).To(gomega.BeNumerically("==", 0))
+			ginkgo.By("Verify their positions and priorities", func() {
+				expectedWls := []visibility.PendingWorkload{
```
Comment (mimowo) [inline test/e2e/e2e_test.go:226]:
```suggestion
				wantPendingWorkloads := []visibility.PendingWorkload{
```
nit

File: apis/visibility/v1alpha1/types.go (hunk 2 above)
Comment (alculquicondor) [reply apis/visibility/v1alpha1/types.go:87]: How so? we use these types in k/k extensively.

What is the conversion for?

**Review by mimowo — COMMENTED** [review]
Just some nits

File: test/e2e/e2e_test.go (hunk 5 above)
Comment (PBundyra) [reply test/e2e/e2e_test.go:252]: The two first workloads have different names and localQueues, since we don't know their order I wanted to avoid checking this. An alternative would be to wait until one the first workloads gets created or add an additional priority class.

**Review by mimowo — COMMENTED** [review]
Generally LGTM, remaining question about the conversion. I think potential changing the types could be handled outside of the this PR.

**Comment by mimowo** [comment]
Oh, I see why the e2e tests are failing, because the workload name is actually derived from the job name according to: https://github.com/kubernetes-sigs/kueue/blob/1ecd79f748af2617111d0bfa1ca848f9e172f4d7/pkg/controller/jobframework/workload_names.go#L44C8-L44C8. Can we call this function from within e2e tests? If not, maybe we need to fetch the workloads (per job), with GET, and assert on the names.

File: apis/visibility/v1alpha1/types.go (hunk 2 above)
Comment (mimowo) [reply apis/visibility/v1alpha1/types.go:87]: > What is the conversion for?

`func autoConvert_url_Values_To_v1alpha1_PendingWorkloadOptions(in *url.Values, out *PendingWorkloadOptions, s conversion.Scope) error {` in `zz_generated.conversion.go`.

Comment (PBundyra) [reply apis/visibility/v1alpha1/types.go]: I've double-checked, and to my surprise, int64 is supported but int32 is not. In general, the conversion is used to convert `url.Values` (query parameters) into `PendingWorkloadOptions`. 

I'll use int64 then.

File: test/e2e/e2e_test.go (hunk 5 above)
Comment (PBundyra) [reply test/e2e/e2e_test.go]: I've added an additional priority class

**Comment by PBundyra** [comment]
> Oh, I see why the e2e tests are failing, because the workload name is actually derived from the job name according to: https://github.com/kubernetes-sigs/kueue/blob/1ecd79f748af2617111d0bfa1ca848f9e172f4d7/pkg/controller/jobframework/workload_names.go#L44C8-L44C8. Can we call this function from within e2e tests? If not, maybe we need to fetch the workloads (per job), with GET, and assert on the names.

I didn't see this comment earlier, and I've suppressed checking the Name. Do you think it's necessary to check it, as we already do it in unit tests?

**Comment by mimowo** [comment]
> I didn't see this comment earlier, and I've suppressed checking the Name. Do you think it's necessary to check it, as we already do it in unit tests?

I think it is ok to skip the check, given that it is done in unit tests + the other properties of the workload identify it in the e2e tests.

File: apis/visibility/v1alpha1/conversion_test.go [hunk 7]
```diff
@@ -46,8 +46,8 @@ func TestPendingWorkloadsOptions(t *testing.T) {
 				"offset": {"2"},
 			},
 			wantPendingWorkloadOptions: PendingWorkloadOptions{
-				Limit:  1,
```
Comment (mimowo) [inline apis/visibility/v1alpha1/conversion_test.go:49]: Is this change required?

File: pkg/constants/constants.go [hunk 8]
```diff
@@ -38,5 +38,5 @@ const (
 	WorkloadPriorityClassSource = "kueue.x-k8s.io/workloadpriorityclass"
 	PodPriorityClassSource      = "scheduling.k8s.io/priorityclass"
 
-	DefaultPendingWorkloadsLimit = 1000
```
Comment (mimowo) [inline pkg/constants/constants.go:41]: Is this required ?

File: apis/visibility/v1alpha1/conversion_test.go (hunk 7 above)
Comment (PBundyra) [reply apis/visibility/v1alpha1/conversion_test.go]: Good point, I've reverted it

File: pkg/constants/constants.go (hunk 8 above)
Comment (PBundyra) [reply pkg/constants/constants.go]: Fixed

File: pkg/visibility/api/rest/pending_workload_CQ_test.go [hunk 9]
```diff
@@ -0,0 +1,301 @@
...
+					PositionInClusterQueue: 1,
+					PositionInLocalQueue:   1,
+				},
+			},
+		},
+	}
+	for name, tc := range cases {
+		t.Run(name, func(t *testing.T) {
+			manager := queue.NewManager(utiltesting.NewFakeClient(), nil)
+			ctx := context.Background()
```
Comment (mimowo) [inline pkg/visibility/api/rest/pending_workload_CQ_test.go:272]: I think we should wrap the ctx with cancel and timeout, and defer cancel(), for example as here (and cleanup the manager): 
https://github.com/kubernetes-sigs/kueue/blob/1ecd79f748af2617111d0bfa1ca848f9e172f4d7/pkg/scheduler/scheduler_test.go#L1108-L1110

File: test/e2e/e2e_test.go [hunk 10]
```diff
@@ -157,6 +167,100 @@ var _ = ginkgo.Describe("Kueue visibility server", func() {
...
+					{
+						JobName:          "lq-b-low-prio",
+						JobPrioClassName: lowPriorityClass.Name,
+						LocalQueueName:   localQueueB.Name,
+					},
+				}
+				for _, jobCase := range jobCases {
+					job := testingjob.MakeJob(jobCase.JobName, ns.Name).
+						Queue(jobCase.LocalQueueName).
+						Image("gcr.io/k8s-staging-perf-tests/sleep:v0.0.3", []string{"1s"}).
```
Comment (mimowo) [inline test/e2e/e2e_test.go:213]: Why "1s" for the runtime? Can we set 0s, or 1ms instead?

**Review by mimowo — COMMENTED** [review]
Looking good, couple more nits

File: pkg/visibility/api/rest/pending_workload_CQ_test.go [hunk 11]
```diff
@@ -270,6 +270,10 @@ func TestPendingWorkloads(t *testing.T) {
 		t.Run(name, func(t *testing.T) {
 			manager := queue.NewManager(utiltesting.NewFakeClient(), nil)
 			ctx := context.Background()
+			ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
```
Comment (mimowo) [inline pkg/visibility/api/rest/pending_workload_CQ_test.go:273]: This timeout seems very short, I'm worried it might fail on loaded environment, It should not hurt to increase the timeout, to say 5s.

File: pkg/visibility/api/rest/pending_workload_CQ.go [hunk 12]
```diff
@@ -83,7 +97,8 @@ func (m *pendingWorkloadsInCqREST) Get(ctx context.Context, name string, opts ru
 func (m *pendingWorkloadsInCqREST) NewGetOptions() (runtime.Object, bool, string) {
 	// If no query parameters were passed the generated defaults function are not executed so it's necessary to set default values here as well
 	return &v1alpha1.PendingWorkloadOptions{
-		Limit: constants.DefaultPendingWorkloadsLimit,
+		Limit:  constants.DefaultPendingWorkloadsLimit,
+		Offset: 0,
```
Comment (mimowo) [inline pkg/visibility/api/rest/pending_workload_CQ.go:101]: Revert if not necessary

**Review by mimowo — COMMENTED** [review]
/lgtm
/assign @tenzen-y @alculquicondor

**Comment by mimowo** [comment]
LGTM. please squash

File: apis/visibility/v1alpha1/types.go [hunk 13]
```diff
@@ -27,6 +27,11 @@ import (
 type PendingWorkload struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Priority               int32  `json:"priority"`
+	LocalQueueName         string `json:"localqueuename"`
+	PositionInClusterQueue int32  `json:"positioninclusterqueue"`
+	PositionInLocalQueue   int32  `json:"positioninlocalqueue"`
```
Comment (tenzen-y) [inline apis/visibility/v1alpha1/types.go:34]:
```suggestion
	Priority               int32  `json:"priority"`
	LocalQueueName         string `json:"localQueueName"`
	PositionInClusterQueue int32  `json:"positionInClusterQueue"`
	PositionInLocalQueue   int32  `json:"positionInLocalQueue"`
```

Can we use camel case as well as other APIs?

File: pkg/queue/manager.go [hunk 14]
```diff
@@ -571,30 +571,34 @@ func (m *Manager) getClusterQueue(cqName string) ClusterQueue {
 	return m.clusterQueues[cqName]
 }
 
-// PendingWorkloadsInfo computes a head of the new snapshot. Length of the head is set by the maxCount parameter
-func (m *Manager) PendingWorkloadsInfo(cqName string) []kueue.ClusterQueuePendingWorkload {
-	pendingWlsInfo := make([]kueue.ClusterQueuePendingWorkload, 0)
+func (m *Manager) GetPendingWorkloadsInfo(cqName string) []*workload.Info {
```
Comment (tenzen-y) [inline pkg/queue/manager.go:574]:
```suggestion
func (m *Manager) PendingWorkloadsInfo(cqName string) []*workload.Info {
```

I would suggest referring to the contribution guide for SIG Scheduling.

- https://github.com/kubernetes/community/blob/master/sig-scheduling/CONTRIBUTING.md
- https://go.dev/doc/effective_go#Getters.

**Review by tenzen-y — COMMENTED** [review]
My first pass. I have yet to see the implementation.

**Comment by mimowo** [comment]
Still LGTM

**Comment by alculquicondor** [comment]
/lgtm canel
/approve
Leaving LGTM to @mimowo once comments are addressed.

File: pkg/queue/manager.go [hunk 15]
```diff
@@ -571,30 +571,34 @@ func (m *Manager) getClusterQueue(cqName string) ClusterQueue {
...
+	}
+	return cq.Snapshot()
+}
+
+// UpdateSnapshot computes the new snapshot and replaces if it differs from the
+// previous version. It returns true if the snapshot was actually updated.
+func (m *Manager) UpdateSnapshot(cqName string, maxCount int32) bool {
+	cq := m.getClusterQueue(cqName)
+	if cq == nil {
+		return false
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/queue/manager.go:587]: Yes, we do it here: https://github.com/kubernetes-sigs/kueue/blob/44ad0d50cacd1830c8ce2e50b1b584636d1e972a/pkg/controller/core/clusterqueue_controller.go#L275. Altough, the new code does not actually use the in-memory stored snapshot, just uses the same function to generate one.

File: apis/visibility/v1alpha1/types.go [hunk 16]
```diff
@@ -27,6 +27,11 @@ import (
 type PendingWorkload struct {
 	metav1.TypeMeta   `json:",inline"`
 	metav1.ObjectMeta `json:"metadata,omitempty"`
```
[the remark this answers is not in the document]
Comment (mimowo) [reply apis/visibility/v1alpha1/types.go:29]: They correspond to the workload object, similarly as for the visibility in status we decided. Otherwise we also need to return the "kind". Keeping the objects small was the motivation for the status visibility, here I think we can afford more fields easily. So we could have a follow up with JobName, and jobKind fields.

[the remark this answers is not in the document]
Comment (alculquicondor) [reply apis/visibility/v1alpha1/types.go:29]: I think the other way makes more sense, since you are already adding TypeMeta here.

[the remark this answers is not in the document]
Comment (mimowo) [reply apis/visibility/v1alpha1/types.go:29]: I see, I think we could try that, but indeed I suggest a follow up, maybe we should update the KEP for it.

I thought the `TypeMeta` here is to indicate details about the `PendingWorkload` type itself, isn't it?

[the remark this answers is not in the document]
Comment (alculquicondor) [reply apis/visibility/v1alpha1/types.go:29]: Right now we don't have an endpoint for an individual Workload or Job, so we are not actually using the TypeMeta right now.
But unclear if we will need a different struct for that.

[the remark this answers is not in the document]
Comment (mimowo) [reply apis/visibility/v1alpha1/types.go:29]: We intend to use the same struct `PendingWorkload` for all three endpoints: clusterqueue, localqueue, and individual workload. 

Just to be clear on the suggestion. `metav1.TypeMeta` has `Kind` and `APIVersion` fields. So, say, for a batch job we would return instance of `PendingWorkload` with `.kind=Job`, `.apiversion=batch/v1`, then `.name={job.name}`. 

I like the suggestion, so that we use the concept of Jobs, which are supplied by users. However, as you mention let's keep it as a follow up. We may also have a small KEP update to clarify this.

[the remark this answers is not in the document]
Comment (PBundyra) [reply apis/visibility/v1alpha1/types.go:29]: To maintain this discussion I'm mentioning [issue#168](https://github.com/kubernetes-sigs/kueue/issues/168) so we do not lose visibility of it.

File: pkg/visibility/api/rest/pending_workload_CQ.go [hunk 17]
```diff
@@ -66,15 +66,29 @@ func (m *pendingWorkloadsInCqREST) Get(ctx context.Context, name string, opts ru
...
-	for index := offset; index < offset+limit && index < int64(len(allPendingWorkloads)); index++ {
-		wlInfo := allPendingWorkloads[index]
-		wls = append(wls, v1alpha1.PendingWorkload{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      wlInfo.Name,
-				Namespace: wlInfo.Namespace,
-			},
-		})
+	pendingWorkloadsInfo := m.queueMgr.PendingWorkloadsInfo(name)
+	localQueuePositions := make(map[string]int32, 0)
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/visibility/api/rest/pending_workload_CQ.go:70]: Note that if LocalQueues are empty we would be over-allocating here. Also, in order to make sure the number of LocalQueues is accurate, then recording their number would be another operation during the read lock. Given  the trade-off is not clear, and would suggest leaving as is.

File: apis/visibility/v1alpha1/types.go [hunk 18]
```diff
@@ -27,6 +27,18 @@ import (
...
+
+	// Priority indicates the workload's priority
+	Priority int32 `json:"priority"`
+
+	// LocalQueueName indicates the name of the LocalQueue the workload is submitted to
+	LocalQueueName string `json:"localQueueName"`
+
+	// PositionInClusterQueue indicates the workload's position in the ClusterQueue starting from 0
```
Comment (mimowo) [inline apis/visibility/v1alpha1/types.go:37]:
```suggestion
	// PositionInClusterQueue indicates the workload's position in the ClusterQueue, starting from 1
```
I'm leaning to "Position" starting from 1 `1`:) , because this is a user-facing API (and "1" seems more natural to the user). I would be leaning towards 0 if this was "Index".

Comment (mimowo) [reply apis/visibility/v1alpha1/types.go]: Indeed we say in the KEP that offset starts from 0: `offset indicates position of the first fetched workload - default: 0`. In that case, I'm ok to leave it as is in this PR, but we could consider renaming in a follow up while the API is still alpha, based on user feedback.

**Comment by mimowo** [comment]
LGTM, please squash

**Comment by mimowo** [comment]
> Leaving LGTM to @mimowo once comments are addressed.

I consider the comments addressed. 
/lgtm

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
Visibility API: extend the information returned for the pending workloads in a ClusterQueue, including the workload position in the queue.
```
